[Elixir] new concept exercise - bitstrings (#1593)
exercism/v3: languages/elixir/exercises/concept/bitstrings/.docs/introduction.md

Working with binary data is an important concept in any language, and Elixir provides an elegant syntax to write, match, and construct binary data.

## Bitstrings

> In Elixir, binary data is referred to as the bitstring type. The binary datatype is a specific form of a bitstring, which we will discuss in a later exercise.

Reviewer: I had to read this sentence a few times to understand it. "binary data" and "binary datatype" sound like the same thing if you're not paying extra attention. Maybe this will make it clearer?

```suggestion
In Elixir, binary data is referred to as the bitstring type. The binary data*type* (not to be confused with binary data in general) is a specific form of a bitstring, which we will discuss in a later exercise.
```

Author: Wonderful!

Bitstring literals are defined using the bitstring special form `<<>>`. When defining a bitstring literal, you must also specify how many bits the value will take up using the `::` type declaration operator.

```elixir
# This defines a bitstring with three segments of a single bit each

> <<0::1, 1::1, 0::1>> == <<0::size(1), 1::size(1), 0::size(1)>>
> <<0::1, 1::1, 0::1>> == <<2::size(3)>>

Reviewer: Maybe let's be explicit?
```suggestion
<<0::1, 1::1, 0::1>> == <<0::size(1), 1::size(1), 0::size(1)>>
# => true
<<0::1, 1::1, 0::1>> == <<2::size(3)>>
# => true
```

```

If the value of the segment overflows the capacity of the segment's type, it will be truncated from the left.

When writing binary integer literals, we can write them directly in base-2 notation by prefixing the literal with `0b`: E.g. `0b10 == 2`

```elixir
value = 0b11111011011
<<value::11>>
# => <<251, 3::size(3)>>
# Note that by default, bitstrings are displays in chunks of 8 bits (a byte)
```

## Constructing

We can combine bitstrings stored in variables using the special form:

```elixir

> value = 0b0110
> first = <<value::4>>
> second = <<0b1001::4>>
> combined = <<first::bitstring, second::bitstring>>

Reviewer: Let's actually show the result of the combination:
```suggestion
first = <<0b110::3>>
second = <<0b001::3>>
combined = <<first::bitstring, second::bitstring>>
# => <<49::size(6)>>
```

I had to change the numbers because the original example returns `"i"`.

<<0::1, 1::1, 0::1>> == <<0::size(1), 1::size(1), 0::size(1)>>
# => true
<<0::1, 1::1, 0::1>> == <<2::size(3)>>
# => true
<<2::1>> == <<0::1>>
# => true because of value overflow
```

When writing binary integer literals, we can write them directly in base-2 notation by prefixing the literal with `0b`: E.g. `0b10 == 2`

```elixir
value = 0b11111011011 = 2011
<<value::11>>
# => <<251, 3::size(3)>>
# By default, bitstrings are displayed in chunks of 8 bits (a byte)
```

## Constructing

We can combine bitstrings stored in variables using the special form:

```elixir
first = <<0b110::3>>
second = <<0b001::3>>
combined = <<first::bitstring, second::bitstring>>
# => <<49::size(6)>>
```

## Pattern matching

Pattern matching can also be done to obtain the value from within the special form:

```elixir
<<value::4, rest::bitstring>> = <<0b01101001::8>>
value == 0b0110
# => true
```

## Tail Call Recursion and Accumulators

When recursing through enumerables [lists, bitstrings, strings], there are often two concerns:

- how much memory is required to store the trail of recursive function calls
- how to build the solution efficiently

> To deal with these concerns an _accumulator_ may be used to pass the state of the function _so far_ until the result reaches the _base case_ and returns the value. An accumulator is a variable that is passed along in addition to the data.

Reviewer: I really like this explanation!

Author: It builds on the excellent explanation in your `recursive` exercise! :tada:
